klippy/extras/sgp40.py

```diff
@@ -0,0 +1,190 @@
+            self.humidity = self._calculate_humidity(self.temp)
+
+
+        try:
+            cmd = (SGP40_CMD["MEASURE_RAW"]
+                   + self._humidity_to_ticks(self.humidity)
+                   + self._temperature_to_ticks(self.temp))
+            value = self._write_command(cmd, read_len=1, delay=.03)
+            self.raw = value[0]
+            self.voc = self.voc_algorithm.process(self.raw)
+        except Exception as e:
+            logging.error(self._log_message(f"Error processing sample: {e}"))
+
+        measured_time = self.reactor.monotonic()
+        print_time = self.i2c.get_mcu().estimated_print_time(measured_time)
+        self._callback(print_time, self.temp)
+        return measured_time + self.report_time
+
+    def _write_command(self, command, delay=0, read_len=0):
+        self.i2c.i2c_write(command)
+
+        if read_len == 0:
+            return None
+
+        # Delay for the command to be processed
+        if delay > 0:
+            self.reactor.pause(self.reactor.monotonic() + delay)
+
+        reply_len = read_len * SGP40_WORD_LEN
+        params = self.i2c.i2c_read([], reply_len)
+
+        if 'response' not in params or len(params['response']) != reply_len:
+            logging.error(self._log_message("Invalid response length from "
+                                            "sensor"))
+            return None
+
+        response = bytearray(params['response'])
+        data = [unpack_from(">H", response, i)[0] for i in range(0, reply_len,
+                SGP40_WORD_LEN)]
+
+        return data
+
+    def _calculate_humidity(self, temp):
+        # Magnus formula for estimating the saturation vapor pressure curve
+        a = 17.62
+        b = 243.12
+        saturation_vapor_pressure = 6.112 * math.exp((a * temp) / (b + temp))
+        actual_vapor_pressure = 6.112 * math.exp( (a * 25) / (b + 25))
+        relative_humidity = ((actual_vapor_pressure / saturation_vapor_pressure)
+                             * 50)
+        return max(0, min(100, relative_humidity))
+
+    def _temperature_to_ticks(self, temperature):
+        ticks = int(round(((temperature + 45) * 65535) / 175)) & 0xFFFF
+        data = [(ticks >> 8) & 0xFF, ticks & 0xFF]
+        crc = self._generate_crc(data)
+
+        return data + [crc]
+
+    def _humidity_to_ticks(self, humidity):
+        ticks = int(round((humidity * 65535) / 100)) & 0xFFFF
+        data = [(ticks >> 8) & 0xFF, ticks & 0xFF]
+        crc = self._generate_crc(data)
+
+        return data + [crc]
+
+    def _generate_crc(self, data):
+        # From SGP40 data sheet
+        crc = 0xFF
+        for byte in data:
+            crc ^= byte
+            for _ in range(8):
+                if crc & 0x80:
+                    crc = (crc << 1) ^ 0x31
+                else:
+                    crc = crc << 1
+        return crc & 0xFF
+
+    def _log_message(self, message):
+        return f"SGP40 {self.name}: {message}"
```

Comment:
Python 2 does not have f-strings.

```suggestion
        return f"SGP40 %s: %s" % (self.name, message)
```